### PR TITLE
Adds support for skills without AppId and password in emulator.

### LINKED ISSUE #4752: I can develop skills in Emulator without an app password (dotnet)
See [parent](https://github.com/microsoft/botframework-sdk/issues/5726)

### PR DESCRIPTION
**Pull request by gabog**
Fixes #4752

- Added new Skill constant to RoleTypes
- Updated BotframeworkHttpCllient to set the Role in the Recipient of the outgoing activity clone to RoleTypes.Skill
- Updated ChannelServiceHanlder to add an anonymous skill AppId claim when auth is disabled.
- Updated AppCredentials.ShouldSetToken() logic to avoid sending a token when the AppId is not set or when it is set to the anonymous skill appId.
- Updated SkillValidation.IsSkillClaim to return true for anonymous skills.
- Updated JwtTokenValidation.AuthenticateRequest to check the Role property in the incoming activity to determine if the request is coming from another bot or the channel. 
- Added AuthenticationConstants.AnonymousSkillAppId constant.
- Added AuthenticationConstants.AnonymousAuthType constant and updated related code to use it instead of using hardcoded "anonymous" 
- Added unit tests for new logic.

## Testing
Try any skill sample from the emulator without using AppIds and Passwords in appsettings.json

### CODE REVIEW

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (EricDahlvang) [inline libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:72]: This seems like an obscure comparison... assuming if RelatesTo is not null means the the activity is a skill.  Is this defined by some underlying protocol behavior?

Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:72]: I agree that this is obscure and that's why we are keeping internal. In the absence of a token, the only way we currently have of knowing if an activity is coming from a channel or another bot is the RelatesTo property. 
We discussed with @Stevenic and skills are the only ones using RelatestTo at the moment. 
We should consider updating the Activity Schema to add a new property to better inform the receiver if this is a skill call or a channel call, once we do that we can update the internal implementation to use the new property (but AuthenticateRequest will not change since the change will be internal)

File: libraries/Microsoft.Bot.Connector/Authentication/AppCredentials.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Connector/Authentication/AppCredentials.cs:262]: Optional: maybe a bit more info for future devs

```suggestion
                // We don't set the token if the AppId is not set, since it means that we are in an un-authenticated scenario
```

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:72]: @EricDahlvang We have a meeting next week with Dan and others to discuss the final way of doing this. We hope we can do all in R11 but we should evaluate the trade-offs in the case we dn't get those changes in. The challenge is, given an incoming activity, determine if I am a skill at the time or not. RelatesTo is the only thing that we reliably set that seems to determine that. Now, with that and assuming the final protocol solution is not ready for R1, we have 2 options: 1) do this, and allow bot developers and composer users to use skills unauthenticated, knowing that it will work on the vast majority of scenarios but there is a bit of risk  or 2) not do this since it is a bit obscure.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:81]: should we define a constant for this anonymous? It's in 3 places.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:75]: consider: Should we define a method that creates this 'un-authenticated claims identity? We have this exact code in one more place. i.e. return SkillsValidation.AnonymousIdentity which is a constant.

File: libraries/Microsoft.Bot.Connector/Authentication/SkillValidation.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Connector/Authentication/SkillValidation.cs:89]: consider: would it be valuable to have this validation built in in a single place? SkillValidation.IsAnonymous(claims)?

File: libraries/Microsoft.Bot.Builder/ChannelServiceHandler.cs
Comment (johnataylor) [inline libraries/Microsoft.Bot.Builder/ChannelServiceHandler.cs:47]: why are we making this protected? going forward we will be deprecating this as it only exists to support the old ChannelService concept.

**Review by johnataylor — CHANGES_REQUESTED** [review]

File: libraries/Microsoft.Bot.Builder/ChannelServiceHandler.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Builder/ChannelServiceHandler.cs:47]: I John, this is already protected in the product today, I just removed the instance variable (small cleanup), we will need to deprecate this property (or do a small breaking change and remove it) once we have the new auth classes.

File: libraries/Microsoft.Bot.Connector/Authentication/AppCredentials.cs
Comment (stevengum) [reply libraries/Microsoft.Bot.Connector/Authentication/AppCredentials.cs:262]: @johnataylor and I still want to remove the TrustServiceUrl concept; by the time the bot gets to `ShouldSetToken()`, it's already processed the request and has decided to send an activity.

@johnataylor's proposal to remove the TrustServiceUrl concept was to have `ShouldSetToken()` always return true.

Implementing this proposal verbatim likely won't hold as `ProcessHttpRequestAsync()` will likely try to fetch a token with null values, leading to a NullReferenceException or something else. But the proposal is elegant and what I believe we should move forward with.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (stevengum) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:72]: I agree with @EricDahlvang after checking the spec. Null-checking a property that has a complex object doesn't seem safe.

I unfortunately don't remember why the SDK populates the RelatesTo field and there weren't any comments in BFHttpClient explaining why it behaves like this. If someone wanted to obfuscate the user data before it was sent to the skill, they either have to mangle a cloned activity before calling the BFHttpClient or subclass the client and remove the logic. This feels like it should have been an opt-in feature.

**Review by stevengum — CHANGES_REQUESTED** [review]
Blocking until the design meeting with @dandriscoll's team occurs next week. After that, don't block the PR on my behalf.....

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:81]: L75 is for the skill cases and it adds an AppId for skills to the IsSkillClaims check passes.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:75]: I am not sure, this is used in Microsoft.Bot.Builder and Microsoft.Bot.Connector, so we will need to add another static method on SkillValidation (or some other class) and make it public. Given that we are not completely sure this is the right approach, I'd rather leave with some internal duplication and not add a new public method that we will need to deprecate if we change our mind.
Thoughts?

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: FYI, after reviewing with Dan and others we decided to use Role instead of ReleatesTo to check if an activity was sent by a skill or not. The changes are in my latest commit to this PR and I also updated the description the PR to summarize the changes I made.

Are we good here now?

File: libraries/Microsoft.Bot.Connector/Authentication/SkillValidation.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/SkillValidation.cs:89]: Maybe in a future rev? I am trying to reduce the new surface area, we can add this helper once we prove this solution works.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (johnataylor) [inline libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:62]: This change also (unfortunately but temporarily) needs to go into ParameterizedBotFrameworkAuthentication.JwtTokenValidation_AuthenticateRequestAsync()
doesn't it?

**Review by johnataylor — CHANGES_REQUESTED** [review]

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:62]: Done, does it look OK now?

File: libraries/Microsoft.Bot.Connector/Authentication/ParameterizedBotFrameworkAuthentication.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Connector/Authentication/ParameterizedBotFrameworkAuthentication.cs:117]: We're doing this in 3 places now. If we change one, we are always going to change the other. Are we sure we don't want to abstract this?

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Connector/Authentication/ParameterizedBotFrameworkAuthentication.cs:117]: If you want to validate this solution in composer and all that is fine, but eventually once we are happy with this solution, we need to make this code centralized. Otherwise hard to prevent future maintainers to change only one of the places and break some scenarios.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: I get that point. However now we went from places to 3 places. You cannot guarantee that a future dev will remember to update all 3 places if they change it. Furthermore, we cannot even guarantee that a code reviewer will remember that there are 3 places that need to be updated. 

If you wnat to keep the internal dupe until we prove it, that's fair, but eventually would love to converge this. LEaving it at your criteria.

**Review by johnataylor — APPROVED** [review]

**Review by carlosscastro — APPROVED** [review]

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs
Comment (gabog) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: Added SkillValidation.CreateAnonymousSkillClaim() helper and tests and updated related code.

**Pull request merged**
